### PR TITLE
add taxonomy service

### PR DESCRIPTION
**Pull request by sorja**

### CODE REVIEW

**Review by minotogna — CHANGES_REQUESTED** [review]
move taxon/taxa related methods to TaxonService

File: src/taxonomy/service.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import { User } from '../auth'
+import { Taxonomy } from './taxonomy'
+
+export interface Job {}
```
Comment (minotogna) [inline src/taxonomy/service.ts:4]: you need to return JobStatus (that is already implemented) as mentioned

File: src/taxonomy/taxonService.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import { Taxon } from 'src/taxonomy/taxon'
+
+export interface TaxonService {
+  // ==== READ
+  count(options: { surveyId: string; taxonomyUuid: string; draft: boolean }): Promise<number>
+
```
Comment (minotogna) [inline src/taxonomy/taxonService.ts:6]: you must have two main methods with optional parameters:
get and getMany

File: src/taxonomy/taxonService.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import { Taxon } from 'src/taxonomy/taxon'
+
+export interface TaxonService {
+  // ==== READ
+  count(options: { surveyId: string; taxonomyUuid: string; draft: boolean }): Promise<number>
```
Comment (minotogna) [inline src/taxonomy/taxonService.ts:5]: draft is not required

File: src/taxonomy/service.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  // ==== READ
+  count(options: { surveyId: string }): Promise<number>
+
+  getMany(options: {
+    draft: boolean
+    limit: number
+    offset: number
+    search: string
+    surveyId: string
+    validate: boolean
```
Comment (minotogna) [inline src/taxonomy/service.ts:19]: draft, limit, offset, search, validate are optional

File: src/taxonomy/service.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+  // ==== UPDATE
+
+  updateTaxonomyProp(options: {
+    user: User
+    surveyId: string
+    taxonomyUuid: string
+    key: string
+    value: any
+  }): Promise<void>
```
Comment (minotogna) [inline src/taxonomy/service.ts:32]: call this update and pass only props of type TaxonomyProps. not key value anymore

File: src/taxonomy/service.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+  importTaxonomy(options: { user: User; surveyId: string; taxonomyUuid: string; tempFilePath: string }): Promise<Job>
```
Comment (minotogna) [inline src/taxonomy/service.ts:34]: this goes to taxa service and will be:
```
  importTaxa(options: { user: User; surveyId: string; taxonomyUuid: string; filePath: string }): Promise<JobStatus>
```

**Review by minotogna — CHANGES_REQUESTED** [review]

**Review by minotogna — CHANGES_REQUESTED** [review]
surveyId anywhere used is of type number

File: src/taxonomy/service.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  getMany(options: {
+    draft?: boolean
+    limit?: number
+    offset?: number
+    search?: string
+    surveyId: number
+    validate?: boolean
+  }): Promise<Array<Taxonomy>>
+
+  get(options: { surveyId: number; taxonomyUuid: string; draft: boolean; validate: boolean }): Promise<Taxonomy>
```
Comment (minotogna) [inline src/taxonomy/service.ts:20]: draft and validate are optional

File: src/taxonomy/taxonService.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import { JobStatus } from 'src/job'
+import { Taxon } from 'src/taxonomy/taxon'
+
+export interface TaxonService {
+  // ==== READ
+  count(options: { surveyId: number; taxonomyUuid: string; draft?: boolean }): Promise<number>
+
+  get(options: { surveyId: string; taxonomyUuid: string; draft: boolean }): Promise<Taxon>
+
+  getMany(options: { surveyId: string; taxonomyUuid: string; draft: boolean }): Promise<Array<Taxon>>
```
Comment (minotogna) [inline src/taxonomy/taxonService.ts:11]: draft is optional

File: src/taxonomy/taxonService.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    draft: boolean
+    limit: number
+    offset: number
+  }): Promise<Array<Taxon>>
+
+  getTaxonVernacularNameByUuid(options: {
+    surveyId: number
+    vernacularNameUuid: string
+    draft: boolean
+  }): Promise<Taxon>
```
Comment (minotogna) [inline src/taxonomy/taxonService.ts:49]: isn't this getManyVernacularNames and returns and Array of VernacularName ?
and why byUuid ?I don't see any uuid in the options

File: src/taxonomy/taxonService.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+  getTaxonByUuid(options: { surveyId: number; taxonUuid: string; draft: boolean }): Promise<Taxon>
```
Comment (minotogna) [inline src/taxonomy/taxonService.ts:51]: this should get. with taxonUuid as option

File: src/taxonomy/taxonService.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    includeUnlUnk: boolean
+  }): Promise<Array<Taxon>>
+
+  getByVernacularNames(options: {
+    surveyId: number
+    taxonomyUuid: string
+    draft: boolean
+    limit: number
+    offset: number
+  }): Promise<Array<Taxon>>
```
Comment (minotogna) [inline src/taxonomy/taxonService.ts:43]: they should all be merged into getMany.
pass an option searchField?:'code'|'scientificName'|'vernacularNames', searchValue?:string.
plus, draft, limit, offset, includeUnlUnk are optional

File: src/taxonomy/taxonService.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
+import { User } from 'src/auth'
+import { JobStatus } from 'src/job'
+import { Taxon } from 'src/taxonomy/taxon'
+
+export interface TaxonService {
+  // ==== READ
+  count(options: { surveyId: number; taxonomyUuid: string; draft?: boolean }): Promise<number>
+
+  get(options: { surveyId: string; taxonomyUuid: string; draft: boolean }): Promise<Taxon>
```
Comment (minotogna) [inline src/taxonomy/taxonService.ts:9]: draft is optional

**Review by minotogna — CHANGES_REQUESTED** [review]

File: src/taxonomy/taxonService.ts [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+  getMany(options: {
+    draft?: boolean
+    limit?: number
+    offset?: number
+    searchField?: 'code' | 'scientificName' | 'vernacularNames'
+    searchValue?: string
+    surveyId: string
+    taxonomyUuid: string
+    uncludeUnlUnk?: boolean
```
Comment (minotogna) [inline src/taxonomy/taxonService.ts:20]: maybe a typo? `unclude`

File: src/taxonomy/taxonService.ts [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+  getManyVernacularNames(options: {
+    draft?: boolean
+    surveyId: number
+    vernacularNameUuid: string
+  }): Promise<Array<VernacularName>>
+
+  // ==== UPDATE
+
+  importTaxa(options: { user: User; surveyId: number; taxonomyUuid: string; filePath: string }): Promise<JobStatus<any>>
```
Comment (minotogna) [inline src/taxonomy/taxonService.ts:31]: let's just call this `import`

**Review by minotogna — CHANGES_REQUESTED** [review]

File: src/taxonomy/taxonService.ts (hunk 13 above)
Comment (sorja) [reply src/taxonomy/taxonService.ts:20]: yes

File: src/taxonomy/taxonService.ts (hunk 14 above)
Comment (sorja) [reply src/taxonomy/taxonService.ts:31]: import is reserved keyword..

**Review by SteRiccio — APPROVED** [review]

File: src/taxonomy/taxonService.ts (hunk 14 above)
Comment (sorja) [reply src/taxonomy/taxonService.ts:31]: Here is more:
https://262.ecma-international.org/11.0/#sec-keywords-and-reserved-words

Comment (sorja) [reply src/taxonomy/taxonService.ts:31]: We can see also `delete` is reserved, I think this can cause problems and should be avoided

**Review by minotogna — APPROVED** [review]

**Pull request merged**